src/Evolution/Systems/ScalarWave/BoundaryCorrections/UpwindPenalty.hpp

```diff
@@ -0,0 +1,279 @@
+ *    \left\{
+ *        \begin{array}{ll}
+ *          \lambda_{\hat{\alpha}} &
+ *            \mathrm{if}\;\lambda_{\hat{\alpha}}\le 0.0 \\
+ *          0 & \mathrm{otherwise}
+ *        \end{array}\right.
+ * \f}
+ *
+ * Note that we have assumed \f$n_i^{\mathrm{ext}}\f$ points in the same
+ * direction as \f$n_i^{\mathrm{int}}\f$. If \f$n_i^{\mathrm{ext}}\f$ points in
+ * the opposite direction the external speeds have their sign flipped and the
+ * \f$\pm\f$ fields and their speeds reverse roles. Specifically, in the code we
+ * have:
+ *
+ * \f{align}{
+ *   D_{\Psi} &= \bar{\lambda}_{v^{\Psi}}^{\mathrm{ext}}
+ *                v^{\mathrm{ext},\Psi}
+ *                - \tilde{\lambda}_{v^{\Psi}}^{\mathrm{int}}
+ *                v^{\mathrm{int},g}, \\
+ *   D_{\Pi}
+ *              &= \frac{1}{2}\left(\bar{\lambda}_{v^+}^{\mathrm{ext}}
+ *                v^{\mathrm{ext},+} +
+ *                \bar{\lambda}_{v^-}^{\mathrm{ext}}
+ *                v^{\mathrm{ext},-}\right)
+ *                + \bar{\lambda}_{v^\Psi}^\mathrm{ext}\gamma_2
+ *                v^{\mathrm{ext},\Psi}
+ *                \notag \\
+ *              &-\frac{1}{2}\left(\tilde{\lambda}_{v^+}^{\mathrm{int}}
+ *                v^{\mathrm{int},+} +
+ *                \tilde{\lambda}_{v^-}^{\mathrm{int}}
+ *                v^{\mathrm{int},-}\right)
+ *                - \tilde{\lambda}_{v^\Psi}^\mathrm{int}\gamma_2
+ *                v^{\mathrm{int},\Psi} , \\
+ *   D_{\Phi_{i}}
+ *              &= \frac{1}{2}\left(\bar{\lambda}_{v^+}^{\mathrm{ext}}
+ *                v^{\mathrm{ext},+}
+ *                - \bar{\lambda}_{v^-}^{\mathrm{ext}}
+ *                v^{\mathrm{ext},-}\right)n_i^{\mathrm{ext}}
+ *                + \bar{\lambda}_{v^0}^{\mathrm{ext}}
+ *                v^{\mathrm{ext},0}_{i}
+ *                \notag \\
+ *              &-
+ *                \frac{1}{2}\left(\tilde{\lambda}_{v^+}^{\mathrm{int}}
+ *                v^{\mathrm{int},+}
+ *                - \tilde{\lambda}_{v^-}^{\mathrm{int}}
+ *                v^{\mathrm{int},-}\right)n_i^{\mathrm{int}}
+ *                - \tilde{\lambda}_{v^0}^{\mathrm{int}}
+ *                v^{\mathrm{int},0}_{i},
+ * \f}
+ *
+ * where
+ *
+ * \f{align}{
+ *   \bar{\lambda}_{\hat{\alpha}} =
+ *    \left\{
+ *        \begin{array}{ll}
+ *          -\lambda_{\hat{\alpha}} &
+ *            \mathrm{if}\;-\lambda_{\hat{\alpha}}\le 0.0 \\
+ *          0 & \mathrm{otherwise}
+ *        \end{array}\right.
+ * \f}
+ */
+template <size_t Dim>
+class UpwindPenalty final : public BoundaryCorrection<Dim> {
+ private:
+  struct NormalTimesVPlus : db::SimpleTag {
+    using type = tnsr::i<DataVector, Dim, Frame::Inertial>;
+  };
+  struct NormalTimesVMinus : db::SimpleTag {
+    using type = tnsr::i<DataVector, Dim, Frame::Inertial>;
+  };
+  struct Gamma2VPsi : db::SimpleTag {
+    using type = Scalar<DataVector>;
+  };
+  struct CharSpeedsTensor : db::SimpleTag {
+    using type = tnsr::i<DataVector, 3, Frame::Inertial>;
```

Comment:
Why use tensor rather than a `std::array`? As far as I understand, the char speed entries don't really have a direct connection with spacetime directions.

Reply:
Yes, but I can't easily store a `std::array` because all these tags are stuck into a `Variables`. I unfortunately am largely stuck with this "use a Tensor as an array" business